algebra/src/curves/tests.rs

```diff
@@ -280,3 +283,338 @@ pub fn curve_tests<G: ProjectiveCurve>() {
+        }
+
+        {
+            let serialized = vec![0; buf_size - 1];
+            let mut cursor = Cursor::new(&serialized[..]);
+            <GroupAffine<P> as CanonicalDeserialize>::deserialize(&mut cursor).unwrap_err();
+        }
+
+        {
+            let mut serialized = vec![0; a.uncompressed_size()];
+            let mut cursor = Cursor::new(&mut serialized[..]);
+            a.serialize_uncompressed(&mut cursor).unwrap();
+
+            let mut cursor = Cursor::new(&serialized[..]);
+            let b = GroupAffine::<P>::deserialize_uncompressed(&mut cursor).unwrap();
+            assert_eq!(a, b);
+        }
+
+        {
+            a.y = -a.y;
+            let mut serialized = vec![0; a.uncompressed_size()];
+            let mut cursor = Cursor::new(&mut serialized[..]);
+            a.serialize_uncompressed(&mut cursor).unwrap();
+            let mut cursor = Cursor::new(&serialized[..]);
+            let b = GroupAffine::<P>::deserialize_uncompressed(&mut cursor).unwrap();
+            assert_eq!(a, b);
+        }
+
+        {
+            let a = GroupAffine::<P>::zero();
+            let mut serialized = vec![0; a.uncompressed_size()];
+            let mut cursor = Cursor::new(&mut serialized[..]);
+            a.serialize_uncompressed(&mut cursor).unwrap();
+            let mut cursor = Cursor::new(&serialized[..]);
+            let b = GroupAffine::<P>::deserialize_uncompressed(&mut cursor).unwrap();
+            assert_eq!(a, b);
+        }
+    }
+}
+
+pub fn edwards_tests<P: TEModelParameters>()
+    where
+        P::BaseField: PrimeField,
+{
+    edwards_curve_serialization_test::<P>();
+    edwards_from_random_bytes::<P>();
+}
+
+pub fn edwards_from_random_bytes<P: TEModelParameters>()
+    where
+        P::BaseField: PrimeField,
+{
+    use crate::curves::models::twisted_edwards_extended::{GroupAffine, GroupProjective};
+    use crate::ToBytes;
+
+    let buf_size = GroupAffine::<P>::zero().serialized_size();
+
+    let rng = &mut thread_rng();
+
+    for _ in 0..ITERATIONS {
+        let a = GroupProjective::<P>::rand(rng).into_affine();
+        {
+            let mut serialized = vec![0; buf_size];
+            let mut cursor = Cursor::new(&mut serialized[..]);
+            CanonicalSerialize::serialize(&a, &mut cursor).unwrap();
+
+            let mut cursor = Cursor::new(&serialized[..]);
+            let p1 = <GroupAffine<P> as CanonicalDeserialize>::deserialize(&mut cursor).unwrap();
+            let p2 = GroupAffine::<P>::from_random_bytes(&serialized).unwrap();
+            assert_eq!(p1, p2);
+        }
+    }
+
+    for _ in 0..ITERATIONS {
+        let biginteger =
+            <<GroupAffine<P> as AffineCurve>::BaseField as PrimeField>::BigInt::rand(rng);
+        let mut bytes = to_bytes![biginteger].unwrap();
+        let mut g = GroupAffine::<P>::from_random_bytes(&bytes);
+        while g.is_none() {
+            bytes.iter_mut().for_each(|i| *i = i.wrapping_sub(1));
```

Comment:
Hunting and pecking with decreasing the x-coordinate until success?
Most likely alright, but where do we have that piece of code from?